.github/workflows/ftl-build.yml

```diff
@@ -15,83 +18,228 @@ on:
+            foo/bar,enable=${{ github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch' }}
       -
         name: Login to DockerHub
-        if: github.event_name != 'pull_request'
+        if: github.event_name != 'pull_request' && github.event_name != 'workflow_dispatch'
         uses: docker/login-action@v2
         with:
           username: ${{ secrets.DOCKERHUB_USER }}
           password: ${{ secrets.DOCKERHUB_PASS }}
       -
         name: Login to GitHub Container Registry
-        if: github.event_name != 'pull_request'
+        if: github.event_name != 'pull_request' && github.event_name != 'workflow_dispatch'
         uses: docker/login-action@v2
         with:
           registry: ghcr.io
           username: ${{ github.repository_owner }}
           password: ${{ secrets.GITHUB_TOKEN }}
       -
-        name: Build
+        # Add support for more platforms with QEMU (optional)
+        # https://github.com/docker/setup-qemu-action
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+        with:
+          platforms: all
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      -
+        name: Build container and test-compile FTL
         uses: docker/build-push-action@v3
         with:
-          context: ftl-build/${{ matrix.ARCH }}/.
+          context: ftl-build/${{ startsWith(matrix.container, 'alpine') && 'alpine' || 'debian' }}
+          platforms: ${{ matrix.platform }}
           push: false
           target: tester
           tags: ${{ steps.meta.outputs.tags }}
           labels: ${{ steps.meta.outputs.labels }}
-          # Caching disabled for now...
-          # cache-from: type=gha,scope=${{ matrix.ARCH }}
-          # cache-to: type=gha,scope=${{ matrix.ARCH }},mode=max
+          build-args: |
+            CONTAINER=${{ matrix.container }}
       -
-        name: Build (all-in)
-        if: matrix.ARCH == 'x86_64'
+        name: Push builder target and push by digest (Docker Hub)
+        id: build_docker
         uses: docker/build-push-action@v3
         with:
-          context: ftl-build/${{ matrix.ARCH }}/.
-          push: false
-          target: tester-all-in
-          tags: ${{ steps.meta.outputs.tags }}
+          context: ftl-build/${{ startsWith(matrix.container, 'alpine') && 'alpine' || 'debian' }}
+          platforms: ${{ matrix.platform }}
+          push: ${{ github.event_name != 'pull_request' && github.event_name != 'workflow_dispatch' }}
+          target: builder
           labels: ${{ steps.meta.outputs.labels }}
+          build-args: |
+            CONTAINER=${{ matrix.container }}
+          outputs: |
+            type=image,name=${{ env.DOCKER_REGISTRY_IMAGE }},push-by-digest=true,name-canonical=true,push=true
       -
-        name: Push
-        if: github.event_name != 'pull_request'
+        name: Push builder target and push by digest (GitHub Container Registry)
```

Comment:
I can't help but feel there is a better way to do this (i.e push to both Docker Hub and GHCR in a single step), but I am at a loss for suggestions. I guess "If it works", then fine, I just would prefer to see less in the way of code duplication if we didn't really _need_ it.

Maybe @crazy-max wouldn't mind sharing his expertise in this area? Would be much appreciated!

Reply:
Seems @DL6ER followed our new docs to distribute build across multiple runners: https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners and looks good :+1:

Reply:
nit: to avoid dedup with Docker Hub and GHCR for some steps in your workflow, a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) would have been nice.